### ISSUE TITLE
[GPU Accelerated] Impossible to convert between the formats supported by the filter 'graph 0 input from stream 0:0' and the filter 'auto_scale_0'

### ISSUE THREAD

**Issue by maxi1134** [AlexxIT/go2rtc#152]
Hey there!
Thanks again for this great software.

I am encountering this new issue on 2/4 of my cameras where the stream simply does not work.
It seems to be due to a transcoding issue between h264 and h264_nvenc.

This seems to be due to the fact that the coder is already h264 on both sides
Perhaps a verification can be done before starting transcodes, instead that directly feeding them into FFMPEG?

Version: Latest Master-hardware tag as of 06-01-2023
Relevant logs:
```
23:46:08.872 DBG [exec] run url="exec:ffmpeg -hide_banner -hwaccel cuda -hwaccel_output_format cuda -extra_hw_frames 2 -allowed_media_types video -fflags nobuffer -flags low_delay -timeout 5000000 -user_agent go2rtc/ffmpeg -rtsp_transport tcp -i rtsp://192.168.50.1:8554/CH001.sdp -vf transpose=1 -c:v h264_nvenc -g 50 -profile:v high -level:v auto -preset:v p2 -tune:v ll -an -user_agent ffmpeg/go2rtc -rtsp_transport tcp -f rtsp rtsp://localhost:8554/1228bb168379b075eb625fdd9458cdc6"
Input #0, rtsp, from 'rtsp://192.168.50.1:8554/CH002.sdp':te=N/A speed=1.01x    speed=   1x
  Metadata:
    title           : streamed by the GeoVision Rtsp Server
    comment         : CH002.sdp
  Duration: N/A, start: 0.278433, bitrate: N/A
  Stream #0:0: Video: h264 (Baseline), yuvj420p(pc, progressive), 448x252, 15 fps, 15 tbr, 90k tbn
Stream mapping:
  Stream #0:0 -> #0:0 (h264 (native) -> h264 (h264_nvenc))
Press [q] to stop, [?] for help
Impossible to convert between the formats supported by the filter 'graph 0 input from stream 0:0' and the filter 'auto_scale_0'
Error reinitializing filters!
Failed to inject frame into filter network: Function not implemented
Error while processing the decoded data for stream #0:0
Conversion failed!
```

Config File:
```
streams:
  front_door_camera_720p:
    # WORKING  - ffmpeg:rtsp://192.168.0.210/h264Preview_01_main#raw=-vf "fps=15,scale=-1:720"#video=h265
    - ffmpeg:rtsp://192.168.0.210/h264Preview_01_main#raw=-vf "fps=15,scale=-1:720" -b:v 1.2M -maxrate 1.2M -bufsize 500k#video=h264 #hardware
  # front_door_camera_frame:
    # - ffmpeg:rtsp://192.168.0.210/h264Preview_01_main#raw=-vf "fps=1,scale=-1:1080"#video=mjpeg
    
  front_door_camera:
    #- ffmpeg -y -vsync 0 -hwaccel cuda -hwaccel_output_format cuda -i:rtsp://192.168.0.210/h264Preview_01_main#raw= -c:a copy -c:v h264_nvenc
     # - exec:ffmpeg -hwaccel cuda -hwaccel_output_format cuda -i rtsp://192.168.0.210/h264Preview_01_main -c copy  -rtsp_transport tcp -use_wallclock_as_timestamps 1 -c:v h264_nvenc -b:v 5M -f rtsp {output}
    #- rtsp://192.168.0.210/h264Preview_01_main
    # - ffmpeg:rtsp://192.168.0.210/h264Preview_01_main#video=mjpeg
    - ffmpeg:rtsp://192.168.0.210/h264Preview_01_main#video=h264#hardware
  front_door_camera_sub:
    - rtsp://192.168.0.210/h264Preview_01_sub #video=h264
    
  patio_camera:
    - ffmpeg:rtsp://192.168.50.10:554/live/main
  patio_camera_sub:
    - ffmpeg:rtsp://192.168.50.10:554/live/sub
    
  faceshot_camera:
    - ffmpeg:rtsp://192.168.50.9:8554/live0#raw=-vf transpose=1#video=h264#hardware
  faceshot_camera_sub:
    - ffmpeg:rtsp://192.168.50.9:8554/live1#raw=-vf transpose=1#video=h264#hardware
    
  back_door_camera:
    - ffmpeg:rtsp://192.168.50.1:8554/CH001.sdp#raw=-vf transpose=1#video=h264#hardware
  back_door_camera_sub:
    - ffmpeg:rtsp://192.168.50.1:8554/CH002.sdp#raw=-vf transpose=1#video=h264#hardware
# ffmpeg:
  # bin: ffmpeg -y -vsync 0 -hwaccel cuvid -c:v h264_cuvid -i                                        # path to ffmpeg binary
log:
  exec: trace
  
```

**Comment by AlexxIT** [comment]
Try to remove transpose. Maybe it helps.

**Comment by maxi1134** [comment]
> Try to remove transpose. Maybe it helps.

I do need to transpose the image, but will test without it nonetheless and return with the results

**Comment by maxi1134** [comment]
It seems to work without #hardware

I'm thinking hat that is the issue.

**Comment by AlexxIT** [comment]
Don't use `#raw`, use `#rotate`

**Comment by maxi1134** [comment]
Thanks!

I tried replacing with #rotate=90 and adding #hardware back
And now get this error when trying it out

```
19:33:35.086 ERR github.com/AlexxIT/go2rtc/cmd/streams/producer.go:51 > error="exec: exit status 1"
19:33:35.086 DBG [exec] run url="exec:ffmpeg -hide_banner -hwaccel cuda -hwaccel_output_format cuda -extra_hw_frames 2 -allowed_media_types video -fflags nobuffer -flags low_delay -timeout 5000000 -user_agent go2rtc/ffmpeg -rtsp_transport tcp -i rtsp://192.168.50.1:8554/CH001.sdp -c:v h264_nvenc -g 50 -profile:v high -level:v auto -preset:v p2 -tune:v ll -an -vf transpose=1 -user_agent ffmpeg/go2rtc -rtsp_transport tcp -f rtsp rtsp://localhost:8554/376904810faab2997b4ffb92f6e266fb"
Stream mapping:
  Stream #0:0 -> #0:0 (h264 (native) -> h264 (h264_nvenc))
Press [q] to stop, [?] for help
Impossible to convert between the formats supported by the filter 'graph 0 input from stream 0:0' and the filter 'auto_scale_0'
Error reinitializing filters!
Failed to inject frame into filter network: Function not implemented
Error while processing the decoded data for stream #0:0
Conversion failed!
19:33:35.175 ERR github.com/AlexxIT/go2rtc/cmd/streams/producer.go:51 > error="exec: exit status 1"

```

**Comment by AlexxIT** [comment]
Show full link in go2rtc.yaml. You can hide private data.

**Comment by maxi1134** [comment]
```
streams:
  front_door_camera_720p:
    # WORKING  - ffmpeg:rtsp://@192.168.0.210/h264Preview_01_main#raw=-vf "fps=15,scale=-1:720"#video=h265
    - ffmpeg:rtsp://@192.168.0.210/h264Preview_01_main#raw=-vf "fps=15,scale=-1:720" -b:v 1.2M -maxrate 1.2M -bufsize 500k#video=h264 #hardware
  # front_door_camera_frame:
    # - ffmpeg:rtsp://@192.168.0.210/h264Preview_01_main#raw=-vf "fps=1,scale=-1:1080"#video=mjpeg
    
  front_door_camera:
    #- ffmpeg -y -vsync 0 -hwaccel cuda -hwaccel_output_format cuda -i:rtsp:/@192.168.0.210/h264Preview_01_main#raw= -c:a copy -c:v h264_nvenc
     # - exec:ffmpeg -hwaccel cuda -hwaccel_output_format cuda -i rtsp://@192.168.0.210/h264Preview_01_main -c copy  -rtsp_transport tcp -use_wallclock_as_timestamps 1 -c:v h264_nvenc -b:v 5M -f rtsp {output}
    #- rtsp://@192.168.0.210/h264Preview_01_main
    # - ffmpeg:rtsp://@192.168.0.210/h264Preview_01_main#video=mjpeg
    - ffmpeg:rtsp://@192.168.0.210/h264Preview_01_main#video=h264 #hardware
  front_door_camera_sub:
    - rtsp://@192.168.0.210/h264Preview_01_sub#video=h264
    
  patio_camera:
    - ffmpeg:rtsp://192.168.50.10:554/live/main
  patio_camera_sub:
    - ffmpeg:rtsp://192.168.50.10:554/live/sub
    
  faceshot_camera:
    # - ffmpeg:rtsp://@192.168.50.9:8554/live0#raw=-vf transpose=1#video=h264 #hardware
    - ffmpeg:rtsp://@192.168.50.9:8554/live0#rotate=90#video=h264#hardware
  faceshot_camera_sub:
   # - ffmpeg:rtsp://@192.168.50.9:8554/live1#raw=-vf transpose=1#video=h264 #hardware
    - ffmpeg:rtsp://@192.168.50.9:8554/live1#rotate=90#video=h264#hardware
    
  back_door_camera:
   # - ffmpeg:rtsp://192.168.50.3:8554/CH001.sdp#raw=-vf transpose=1#video=h264 #hardware
    - ffmpeg:rtsp://192.168.50.3:8554/CH001.sdp#rotate=90#video=h264#hardware
  back_door_camera_sub:
   # - ffmpeg:rtsp://192.168.50.3:8554/CH002.sdp#raw=-vf transpose=1#video=h264 #hardware
    - ffmpeg:rtsp://192.168.50.3:8554/CH002.sdp#rotate=90#video=h264#hardware
# ffmpeg:yyyyyyy
  # bin: ffmpeg -y -vsync 0 -hwaccel cuvid -c:v h264_cuvid -i                                        # path to ffmpeg binary
log:
  exec: trace
   ````
   
   Does not work, but the folowing works and even spawns ffmpeg processes on my nvidia-smi.
   
   ```
   streams:
  front_door_camera_720p:
    # WORKING  - ffmpeg:rtsp://-@192.168.0.210/h264Preview_01_main#raw=-vf "fps=15,scale=-1:720"#video=h265
    - ffmpeg:rtsp://-@192.168.0.210/h264Preview_01_main#raw=-vf "fps=15,scale=-1:720" -b:v 1.2M -maxrate 1.2M -bufsize 500k#video=h264 #hardware
  # front_door_camera_frame:
    # - ffmpeg:rtsp://-@192.168.0.210/h264Preview_01_main#raw=-vf "fps=1,scale=-1:1080"#video=mjpeg
    
  front_door_camera:
    #- ffmpeg -y -vsync 0 -hwaccel cuda -hwaccel_output_format cuda -i:rtsp://-@192.168.0.210/h264Preview_01_main#raw= -c:a copy -c:v h264_nvenc
     # - exec:ffmpeg -hwaccel cuda -hwaccel_output_format cuda -i rtsp://-@192.168.0.210/h264Preview_01_main -c copy  -rtsp_transport tcp -use_wallclock_as_timestamps 1 -c:v h264_nvenc -b:v 5M -f rtsp {output}
    #- rtsp://-@192.168.0.210/h264Preview_01_main
    # - ffmpeg:rtsp://-@192.168.0.210/h264Preview_01_main#video=mjpeg
    - ffmpeg:rtsp://-@192.168.0.210/h264Preview_01_main#video=h264 #hardware
  front_door_camera_sub:
    - rtsp://-@192.168.0.210/h264Preview_01_sub#video=h264
    
  patio_camera:
    - ffmpeg:rtsp://192.168.50.10:554/live/main
  patio_camera_sub:
    - ffmpeg:rtsp://192.168.50.10:554/live/sub
    
  faceshot_camera:
    # - ffmpeg:rtsp://-@192.168.50.9:8554/live0#raw=-vf transpose=1#video=h264 #hardware
    - ffmpeg:rtsp://-@192.168.50.9:8554/live0#rotate=90#video=h264 #hardware
  faceshot_camera_sub:
   # - ffmpeg:rtsp://-@192.168.50.9:8554/live1#raw=-vf transpose=1#video=h264 #hardware
    - ffmpeg:rtsp://-@192.168.50.9:8554/live1#rotate=90#video=h264 #hardware
    
  back_door_camera:
   # - ffmpeg:rtsp://192.168.50.3:8554/CH001.sdp#raw=-vf transpose=1#video=h264 #hardware
    - ffmpeg:rtsp://192.168.50.3:8554/CH001.sdp#rotate=90#video=h264 #hardware
  back_door_camera_sub:
   # - ffmpeg:rtsp://192.168.50.3:8554/CH002.sdp#raw=-vf transpose=1#video=h264 #hardware
    - ffmpeg:rtsp://192.168.50.3:8554/CH002.sdp#rotate=90#video=h264 #hardware
# ffmpeg:yyyyyyy
  # bin: ffmpeg -y -vsync 0 -hwaccel cuvid -c:v h264_cuvid -i                                        # path to ffmpeg binary
log:
  exec: trace
  
  
 ```

**Comment by severnt** [comment]
I ran into the same issue and came up with a solution.

This returns the same error as in this issue:
```
  streams:
    front: "ffmpeg:rtsp://x:x192.168.100.1:554/stream1#video=h264#rotate=90#hardware"
```

This is because `hardware` causes go2rtc to emit `-hwaccel cuda -hwaccel_output_format cuda`, which works great if you're doing a straight transcode because as I understand it, in this case, both the decode and encode happen on the video card bypassing the CPU and ffmpeg. But if you want to do a rotate, we either need a version of ffmpeg built with `transponse_npp` (which can't be done in a redistributable build because of licensing issues: https://github.com/BtbN/FFmpeg-Builds/issues/50), or we need to tell ffmpeg to retrieve the frames, transform them, then send the results back to the gpu - which is the solution I've gone with.

``` 
  ffmpeg:
    global: "-hide_banner -hwaccel cuda -hwaccel_output_format nv12 -extra_hw_frames 2"
  streams:
    front: "ffmpeg:rtsp://x:x192.168.100.1:554/stream1#video=h264/cuda#rotate=90"
```

This skips go2rtc's hardware detection, sets the hwaccel output to be `nv12` (as opposed to `cuda`), then uses go2rtc's h264/cuda profile (h264_cuvid codec) to re-encode it, and allows for all the CPU based transformations to work.

**Comment by NickM-27** [comment]
You can still use cuda to decode and encode the video, you would just need to set the cuda args in the input but instead of using transpose_npp, use standard transpose. The issue is just that it doesn't seem cuda supports rotation. Go2rtc works fine when doing scaling with cuda_scale

**Comment by AlexxIT** [comment]
Oh. You right. It's impossible to use rotation with any hardware acceleration. You should convert pixel from software to hardware. But it will cost high CPU.

**Comment by aczekajski** [comment]
I had the same error and the weirdest thing helped in my case.
I run it with `-threads 12` and suddenly it works. For everything from 13 on, it fails (my cpu has 16 threads). I've seen someone on the internet having to use 9 so it likely depends on your cpu.

It would be great if someone could explain what is happening here ^^'

**Comment by AlexxIT** [comment]
Latest version was reworked. If hardware doesn't support scaling or rotation - software method will be used.
https://github.com/AlexxIT/go2rtc/releases/tag/v1.6.0
